src/parser.ts

```diff
@@ -53,71 +55,94 @@ function parse(input: string) {
   }
 
   function node() {
+    const root = Tree();
+    let head = root;
+
     expect('[');
     accept_ws();
     if (tok_peek.kind == 'WORD') {
-      nodeType();
+      root.nodeType = nodeType();
       while (accept('.')) {
-        nodeType();
+        const newHead = Tree();
+        head.children.push(newHead);
+        newHead.nodeType = nodeType();
+        head = newHead;
       }
 
       // + '' is to make typescript be quite...
       switch (tok_peek.kind as TokenKind) {
         case '[':
-          nodeList();
+          head.children = nodeList();
           break;
         case '*':
         case '.':
         case '/':
         case '_':
         case '^':
         case 'WORD':
-          nodeData();
+          head.leaf = nodeData();
           break;
       }
     }
 
     expect(']');
+
+    return root;
   }
 
   function nodeType() {
-    expect('WORD');
+    const nt = { name: '', sub: '', sup: '' };
+    nt.name = expect('WORD').value as string;
 
     if (accept('_')) {
-      expect('WORD');
+      nt.sub = expect('WORD').value as string;
```

Comment:
I think if you type the expect function return, you won't have to do type assertion here. But I might be wrong.

Reply:
Actually, now that I'm actually thinking what I MEANT was `nt.sub = epxect('WORD')!.value`. The problem is value is optional.
I am actually going to leave it as is, though. I've had a change of heart about the tokens, and am hoping to make a big update to the token types once a stable V1 is finished.

Reply:
Also, I'm too lazy to merge everything up through these other PRs :p